cpp/src/toppra/algorithm.cpp

```diff
@@ -0,0 +1,80 @@
+#include <cstddef>
+#include <iostream>
+#include <toppra/algorithm.hpp>
+#include <toppra/solver/qpOASES-wrapper.hpp>
+#include "toppra/toppra.hpp"
+
+namespace toppra {
+
+PathParametrizationAlgorithm::PathParametrizationAlgorithm(
+    LinearConstraintPtrs constraints, const GeometricPath &path)
+    : m_constraints(std::move(constraints)), m_path(path){};
+
+ReturnCode PathParametrizationAlgorithm::computePathParametrization(double vel_start,
+                                                                    double vel_end) {
+  ReturnCode ret;
+  initialize();
+  m_solver->setupSolver();
+  Bound vel_ends;
+  vel_ends << vel_end, vel_end;
+  ret = computeControllableSets(vel_ends);
+  if ((int)ret > 0) {
+    return ret;
+  }
+  ret = computeForwardPass(vel_start);
+  return ret;
+};
+
+ReturnCode PathParametrizationAlgorithm::computeControllableSets(
+    const Bound &vel_ends) {
+  TOPPRA_LOG_DEBUG("computeControllableSets");
+  ReturnCode ret = ReturnCode::OK;
+  bool solver_ret;
+  Vector g_upper{2}, g_lower{2}, solution;
+  g_upper << 1e-9, -1;
+  g_lower << -1e-9, 1;
```

Comment:
what is this 1e-9 for ?

Reply:
It's the value used in the Python implementation. I vaguely remember that it's used to break some tie, but couldn't remember exactly why.